### PR TITLE
fix(lsp): robust rename()

### PR DESCRIPTION
**Pull request by tomtomjhj**
* Match the prefix of the list of the path components.
* Directly rename the existing buffers without deleting them, so that states like undo, window layout, etc are preserved.
* Don't write if not writable.
* buffer:// was removed in 236c20795eb9f11e21e0719b735ea741711acc08.
* Remove redundant custom bufwinid function.
* Fix tests.

### CODE REVIEW

File: runtime/lua/vim/lsp/util.lua [hunk 1]
```diff
@@ -771,72 +771,97 @@ function M.text_document_completion_list_to_complete_items(result, prefix)
...
   end
   return buffers
 end
 
+local function escape_gsub_pat(s)
+  local result, _ = s:gsub('(%p)', '%%%1')
+  return result
+end
+
+local function escape_gsub_repl(s)
```
Comment (glepnir) [inline runtime/lua/vim/lsp/util.lua:806]: does there can't use `vim.fn.escape` ?

Comment (tomtomjhj) [reply runtime/lua/vim/lsp/util.lua:806]: lua-pattern and gsub {repl} are escaped with `%`, but escape() uses `\`

Comment (zeertzjq) [reply runtime/lua/vim/lsp/util.lua:806]: Does `vim.psub()` not work?

File: runtime/lua/vim/lsp/util.lua [hunk 2]
```diff
@@ -771,72 +771,97 @@ function M.text_document_completion_list_to_complete_items(result, prefix)
...
-    if bufname:find(path) then
+    local path = path_components(vim.api.nvim_buf_get_name(v))
+    if path_under_prefix(path, prefix) then
       table.insert(buffers, v)
     end
   end
   return buffers
 end
 
+local function escape_gsub_pat(s)
```
Comment (zeertzjq) [inline runtime/lua/vim/lsp/util.lua:801]: `vim.pesc()`

File: runtime/lua/vim/lsp/util.lua (hunk 1 above)
Comment (glepnir) [reply runtime/lua/vim/lsp/util.lua:806]: forget `vim.pesc` then.

**Review by mfussenegger — COMMENTED** [review]
These changes make the behavior actually worse for me. If I rename a class with eclipse.jdt.ls I am afterwards left with error diagnostics because of missing document change notifications.

Could you explain a bit more what problem you're trying to solve here?

**Comment by tomtomjhj** [comment]
The old implementation deletes the buffers that are affected by renaming. Then you will lose the undo information for that buffer, and the windows displaying that buffer will be closed (among many other side effects). I'd like to retain these states even after renaming. 

If you think this is useful, I'll try to fix the missing notification problem. Otherwise, I'll just remove the "Directly rename the existing buffers without deleting them" part of the change.

**Comment by tomtomjhj** [comment]
Oh there's another problem with this change. After renaming, `:write`-ing in the renamed buffer raises E13.

**Comment by tomtomjhj** [comment]
Fixed E13 issue. 

> If I rename a class with eclipse.jdt.ls I am afterwards left with error diagnostics because of missing document change notifications.

@mfussenegger Do you disable didChangeWatchedFiles (as discussed in #23291)? With this enabled, at least rust-analyzer seems to pick up the changes automatically.

**Comment by mfussenegger** [comment]
>  Do you disable didChangeWatchedFiles

No I have it enabled.

Based on the specification I'd think it's important that the `didClose` / `didOpen` and other document synchronization messages happen. If this PR prevents that I don't think we can merge it.

**Comment by tomtomjhj** [comment]
I have modified the implementation to utilize the existing `didClose`/`didOpen` logic for `:saveas`.

**Comment by przepompownia** [comment]
@tomtomjhj could you to merge the current master here? 

Currently on this branch after calling `vim.lsp.util.rename('src/Old.php', 'src/New.php')` both files exists and `Old.php` remains the current buffer in the window (like before call). 

I ended up here by #27291 but am also interested in what it does in practice. I try to debug why the index of phpactor LSP is outdated after renaming a class. If I make this call on master nvim, I see four client notifications:

```log
"rpc.send" { jsonrpc = "2.0", method = "textDocument/didSave", params = { textDocument = { uri = "file:///.../src/Old.php" } }}
"rpc.send" { jsonrpc = "2.0", method = "textDocument/didOpen", params = { textDocument = { languageId = "php", text = "<?php\n\nnamespace ... \n", uri = "file:///.../src/New.php", version = 0 } }}
"rpc.send" { jsonrpc = "2.0", method = "textDocument/didClose", params = { textDocument = { uri = "file:///.../src/Old.php" } }}
"rpc.send" { jsonrpc = "2.0", method = "workspace/didChangeWatchedFiles", params = { changes = { { type = 3, uri = "file:///.../src/Old.php" }, { type = 1, uri = "file:///.../src/New.php" } } }}

```
Is there `textDocument/didSave` for the new file missing?

**Comment by tomtomjhj** [comment]
rebased

> Currently on this branch after calling `vim.lsp.util.rename('src/Old.php', 'src/New.php')` both files exists and `Old.php` remains the current buffer in the window (like before call).

the issue seems to be that it works only with normalized absolute path

**Comment by tomtomjhj** [comment]
> the issue seems to be that it works only with normalized absolute path

fixed

File: runtime/lua/vim/lsp/util.lua [hunk 3]
```diff
@@ -639,76 +639,108 @@ function M.text_document_completion_list_to_complete_items(result, prefix)
...
 function M.rename(old_fname, new_fname, opts)
   opts = opts or {}
-  local target_exists = uv.fs_stat(new_fname) ~= nil
-  if target_exists and not opts.overwrite or opts.ignoreIfExists then
-    vim.notify('Rename target already exists. Skipping rename.')
+  local skip = not opts.overwrite or opts.ignoreIfExists
+
+  local old_fname_full = vim.uv.fs_realpath(vim.fs.normalize(old_fname, { expand_env = false }))
+  if not old_fname_full then
+    vim.notify('Invalid path: ' .. old_fname)
```
Comment (przepompownia) [inline runtime/lua/vim/lsp/util.lua:687]:
```suggestion
    vim.notify('Invalid path: ' .. old_fname, vim.log.levels.ERROR)
```

**Comment by przepompownia** [comment]
With the recent change I get 
```log
"rpc.send" { jsonrpc = "2.0", method = "textDocument/didClose", params = { textDocument = { uri = "file:///.../src/Old.php" } }}
"rpc.send" { jsonrpc = "2.0", method = "textDocument/didOpen", params = { textDocument = { languageId = "php", text = "<?php\n\nnamespace ...\n", uri = "file:///.../src/New.php", version = 0 } }}
"rpc.send" { jsonrpc = "2.0", method = "textDocument/didSave", params = { textDocument = { uri = "file:///.../src/New.php" } }}
"rpc.send" { jsonrpc = "2.0", method = "workspace/didChangeWatchedFiles", params = { changes = { { type = 3, uri = "file:///.../src/Old.php" }, { type = 2, uri = "file:///.../src/New.php" } } }}
```
I didn't check yet if it (especially `didSave` notification for the new file) can fix non-trivial rename operation in phpactor.

**Comment by przepompownia** [comment]
Probably using `:saveas` here solves the same problem as in #27408: the new buffer becomes visible on buffer list (`:ls` without `!`) and I can alternate it with another buffer. Consider tests it from like in #27408 or another way.

**Review by mfussenegger — COMMENTED** [review]
I'm sorry but I still don't really understand the problem(s?) that this PR solves.

I'd rather go with more targeted and smaller PRs like https://github.com/neovim/neovim/pull/27408 instead which I can reason about more easily.

**Comment by przepompownia** [comment]
Nice work, worth testing some of the cases described in individual commits. 

> I have split the commits so that each of them fixes one problem.

**Comment by tomtomjhj** [comment]
I think now the test covers most of the important cases in the `fix` commits.

**Comment by tomtomjhj** [comment]
I'll make prs for smaller units of changes

**Pull request closed without merging**